### ISSUE TITLE
JwtClaimTypes.AuthenticationTime and JwtClaimTypes.IdentityProvider doesn't exist for my external id_tokens

### ISSUE THREAD

**Issue by ghstahl** [DuendeSoftware/IdentityServer#51]
**Which exact version are you using?**
5.0.0-preview.2.13

**Describe the bug**
I am using IDS4 as an OAuth2 token service where I am doing exchanges of external id_tokens (google) for access_tokens, etc.
No id_tokens being produced by IDS4.

I am doing devicecode update where my  [id_token](https://jwt.io/#debugger-io?token=eyJhbGciOiJSUzI1NiIsImtpZCI6ImUxOTdiZjJlODdiZDE5MDU1NzVmOWI2ZTVlYjQyNmVkYTVkNTc0ZTMiLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiJodHRwczovL2FjY291bnRzLmdvb2dsZS5jb20iLCJhenAiOiIxMDk2MzAxNjE2NTQ2LWVkYmw2MTI4ODF0N3JrcGxqcDNxYTNqdW1pbnNrdWxvLmFwcHMuZ29vZ2xldXNlcmNvbnRlbnQuY29tIiwiYXVkIjoiMTA5NjMwMTYxNjU0Ni1lZGJsNjEyODgxdDdya3BsanAzcWEzanVtaW5za3Vsby5hcHBzLmdvb2dsZXVzZXJjb250ZW50LmNvbSIsInN1YiI6IjEwNDc1ODkyNDQyODAzNjY2Mzk1MSIsImF0X2hhc2giOiJzRUlIUXZsMWtuU213N3cxMEdxU0hBIiwibm9uY2UiOiI2Mzc0MzgzMDk3MDEzNzY1OTYuTldJMk9XUTBaR1V0TlRWa1pTMDBaakF5TFRsbFl6VXRaV0k0WW1RMlpHWXpaVEptTVRJeE9UVTNNbVl0WkRjNU5pMDBaVEl3TFRsa056TXROakJsTXpNek9UbGlPVFZtIiwibmFtZSI6IkhlcmIgU3RhaGwiLCJwaWN0dXJlIjoiaHR0cHM6Ly9saDMuZ29vZ2xldXNlcmNvbnRlbnQuY29tL2EtL0FPaDE0R2dPMzZKQURuTWpfXzJqWE4zRk1CelFBOGpITVJpOVVXUVpvcUNUd0E9czk2LWMiLCJnaXZlbl9uYW1lIjoiSGVyYiIsImZhbWlseV9uYW1lIjoiU3RhaGwiLCJsb2NhbGUiOiJlbiIsImlhdCI6MTYwODIzNDE3NCwiZXhwIjoxNjA4MjM3Nzc0fQ.dTcAWg2uqtx7g3MApfN6HJQo1dKNpRG6w9OEllQkVrxnCIosQ4ClDYnYgVgwS55ybuCV94DHwwGW_EDouHJq5bWLcyviB0cuzkLyZVwjw6GAk1mNE0afgMeAMmTgtj9uWUYJxteEp_n80fJYg0259VhsoSz_2ndUaiszzWqIg3PQM21HNzVfQn9B05v2zEWxNjkSEmsDwBLqX9j2Vod03SFwV0R1REozWzgBPS7s8PB2kzILh8IvtxivJ9FgEaM7gKhx3M4zcHfNLzI3I5IPOcvH0Tojm7l0WZQ93PKEUop2cLZ_xf4Hw39CmIqXzYaSGlfxMJp-D5EJW-7tttSYnw&publicKey=-----BEGIN%20PUBLIC%20KEY-----%0AMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAt%2BEAePKVyMaQbjG96EP9%0A8IIB7%2BCJLeo4AZOdc1WVTfMZgdbN%2FcsbY1WP25CrlXvhaIewRNEKTF9WkKGvsxow%0ApYJ%2F18rtOYnz94mn9s%2FEvJaBtoEcixXedwMwniw78ayLyi4IGzCLhUopgLnwAFar%0Adde9ZxpEAVqMK3q4EdScMLebrdrTu63oZ2EpLLIvuC5tjitFXLtNb5v2yiOElX3n%0AXntOF9OYTtpCRzKRVOZ1Lqcj7G3oWmmBmLrR%2BfRc5yFpLFRVHu%2Bvdp4BGUh96t2f%0Alz95QxhIRF0zcuvRiCPWjdiRZgJ8wiSy627XeINqKaoVycW0TofFcz2xAix9GuNd%0AqQIDAQAB%0A-----END%20PUBLIC%20KEY-----%0A) is from google. 

I make a ClaimsPrincipal our of that id_token and it doesn't contain JwtClaimTypes.AuthenticationTime or JwtClaimTypes.IdentityProvider.

```
deviceAuth.Subject = googlePrincipal;
await _deviceFlowStore.UpdateByUserCodeAsync(data.UserCode, deviceAuth);
```

Later in the flow a call is made to the token endpoint with the grant_type:```urn:ietf:params:oauth:grant-type:device_code```.
Which eventually hits the ```GetStandardSubjectClaims``` method, which throws because there is no JwtClaimTypes.AuthenticationTime.

```
protected virtual IEnumerable<Claim> GetStandardSubjectClaims(ClaimsPrincipal subject)
{
    var claims = new List<Claim>
    {
        new Claim(JwtClaimTypes.Subject, subject.GetSubjectId()),
        new Claim(JwtClaimTypes.AuthenticationTime, subject.GetAuthenticationTimeEpoch().ToString(), ClaimValueTypes.Integer64),
        new Claim(JwtClaimTypes.IdentityProvider, subject.GetIdentityProvider())
    };

    claims.AddRange(subject.GetAuthenticationMethods());

    return claims;
}
```

I can create a ClaimsPincipal with it in there by copying it from google's id_token ```iat```, however before doing that I need to know if this is a defect on google's side by not having it or IDS4 assuming that its there because of how IDS4 creates id_tokens.

Is this on me to make sure those 2 claims are in there or something IDS4 accounts for NOT being in there?

```
System.InvalidOperationException: auth_time is missing.
   at Duende.IdentityServer.Extensions.PrincipalExtensions.GetAuthenticationTimeEpoch(IIdentity identity) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\Extensions\PrincipalExtensions.cs:line 52
   at Duende.IdentityServer.Extensions.PrincipalExtensions.GetAuthenticationTimeEpoch(IPrincipal principal) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\Extensions\PrincipalExtensions.cs:line 38
   at Duende.IdentityServer.Services.DefaultClaimsService.GetStandardSubjectClaims(ClaimsPrincipal subject) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\Services\Default\DefaultClaimsService.cs:line 226
   at Duende.IdentityServer.Services.DefaultClaimsService.GetAccessTokenClaimsAsync(ClaimsPrincipal subject, ResourceValidationResult resourceResult, ValidatedRequest request) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\Services\Default\DefaultClaimsService.cs:line 165
   at FluffyBunny4.Services.MyDefaultClaimsService.GetAccessTokenClaimsAsync(ClaimsPrincipal subject, ResourceValidationResult resourceResult, ValidatedRequest request) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Apps\FluffyBunny4\Services\Default\MyDefaultClaimsService.cs:line 42
   at Duende.IdentityServer.Services.DefaultTokenService.CreateAccessTokenAsync(TokenCreationRequest request) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\Services\Default\DefaultTokenService.cs:line 187
   at FluffyBunny4.Stores.MyTokenResponseGenerator.CreateAccessTokenAsync(ValidatedTokenRequest request) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Apps\FluffyBunny4\Stores\MyTokenResponseGenerator.cs:line 115
   at Duende.IdentityServer.ResponseHandling.TokenResponseGenerator.ProcessDeviceCodeRequestAsync(TokenRequestValidationResult request) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\ResponseHandling\Default\TokenResponseGenerator.cs:line 261
   at Duende.IdentityServer.ResponseHandling.TokenResponseGenerator.ProcessAsync(TokenRequestValidationResult request) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\ResponseHandling\Default\TokenResponseGenerator.cs:line 99
   at Duende.IdentityServer.Endpoints.TokenEndpoint.ProcessTokenRequestAsync(HttpContext context) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\Endpoints\TokenEndpoint.cs:line 98
   at Duende.IdentityServer.Endpoints.TokenEndpoint.ProcessAsync(HttpContext context) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\Endpoints\TokenEndpoint.cs:line 70
   at Duende.IdentityServer.Hosting.IdentityServerMiddleware.Invoke(HttpContext context, IEndpointRouter router, IUserSession session, IEventService events, IBackChannelLogoutService backChannelLogoutService) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\Hosting\IdentityServerMiddleware.cs:line 74
   at Duende.IdentityServer.Hosting.IdentityServerMiddleware.Invoke(HttpContext context, IEndpointRouter router, IUserSession session, IEventService events, IBackChannelLogoutService backChannelLogoutService) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\Hosting\IdentityServerMiddleware.cs:line 89
   at Duende.IdentityServer.Hosting.MutualTlsEndpointMiddleware.Invoke(HttpContext context, IAuthenticationSchemeProvider schemes) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\Hosting\MutualTlsEndpointMiddleware.cs:line 94
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Duende.IdentityServer.Hosting.BaseUrlMiddleware.Invoke(HttpContext context) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Duende\Main\IdentityServer\Hosting\BaseUrlMiddleware.cs:line 31
   at FluffyBunny4.Middleware.TenantMiddleware.Invoke(HttpContext context, IScopedTenantRequestContext scopedTenantRequestContext, ITenantStore tenantStore) in C:\work\github\DuendeSoftware-FluffyBunny\TokenServices\src\Apps\FluffyBunny4\Middleware\TenantMiddleware.cs:line 71
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

**Comment by leastprivilege** [comment]
tbh - I don't follow. Could you explain? ;)

**Comment by ghstahl** [comment]
Summary: For me I have talked myself into this not being an IDS4 bug.
I have decided to make sure that auth_time and IDP is present in my ClaimsPrincipal using devicecode flow.

The following is a more step by step of my reasoning.

Ok, so I want to try out the devicecode flow.

1. Make a call to IDS4 deviceauthorization endpoint - works  
I get a user_code, device_code, etc.

2. using backend code I look up the DeviceCode using the UserCode - works
var deviceAuth = await _deviceFlowStore.FindByUserCode(user_code);

3. I then create a ClaimsPrincipal out of a google id_token - works
However I notice that the ClaimsPrincipal doesn't have auth_time or IDP 

4.  I set the  Subject in deviceAuth  to the google ClaimsPrincipal
deviceAuth.Subject = googleClaimsPrincipal;

5. I update the deviceCode;
await _deviceFlowStore.UpdateByUserCodeAsync(data.UserCode, deviceAuth);

6. The client is polling using the device_code from step one to redeem a token
This throws an exception.  The cause is that GetStandardSubjectClaims is called and it expects that auth_time and idp is in the claims principal.

Fix Options:
1. When I create a claimsPrincipal from that google id_token, put in auth_time and IDP so that redeeming the device code works
```
                var claims = validatedResult.User.Claims.ToList();
                claims.Add(new Claim(JwtClaimTypes.AuthenticationTime,
                    claims.FirstOrDefault(c => c.Type == JwtClaimTypes.IssuedAt).Value));
                claims.Add(new Claim(JwtClaimTypes.IdentityProvider, _tokenExchangeOptions.AuthorityKey));

                var newClaimsIdentity = new ClaimsIdentity(claims);
                var subjectPrincipal = new ClaimsPrincipal(newClaimsIdentity);
                
                deviceAuth.Subject = subjectPrincipal;

```
2. Change IDS4 to not require that auth_time and IDP be in the claimsPrincipal.

Should  auth_time and IDP be required to be in the claimsPrincipal for devicecode flow to work.  


When creating a ClaimsPrincipal from an id_token adding the IDP where one doesn't exist in the id_token doesn't change the data of what the id_token is stating.   Adding the IDP makes the ClaimsPrincipal more accurate than the id_token.

However adding an auth_time where one did not exit in the id_token is not being truthful to the claims the id_token is making.  At least when I did it I used the ```iat``` in the id_token as my auth_time, which is at best a half-truth.  


Another question:  The ```GetStandardSubjectClaims``` is saying that the following must be present;
```
JwtClaimTypes.Subject
JwtClaimTypes.AuthenticationTime
JwtClaimTypes.IdentityProvider
```
The google id_token says NOPE, not me.
google doesn't send auth_time and I tried prompt=login and max_age=0


Who is right?
